c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +234,331 @@ struct Block {
+      if (status == CUDA_ERROR_OUT_OF_MEMORY) {
+        for (auto j : c10::irange(begin, i)) {
+          auto handle = handles_.at(j).value();
+          handles_.at(j) = c10::nullopt;
+          C10_CUDA_DRIVER_CHECK(cuMemRelease(handle));
+        }
+        trimHandles();
+        return rangeFromHandles(begin, begin);
+      }
+      C10_CUDA_DRIVER_CHECK(status);
+      handles_.at(i) = handle;
+    }
+    for (auto i : c10::irange(begin, end)) {
+      C10_CUDA_DRIVER_CHECK(cuMemMap(
+          ptr_ + i * segment_size_,
+          segment_size_,
+          0,
+          handles_.at(i).value(),
+          0ULL));
+    }
+
+    setAccess(device_, begin, end);
+    for (auto p : peers_) {
+      setAccess(p, begin, end);
+    }
+    return rangeFromHandles(begin, end);
+  }
+
+  // unmaps all the completely empty segment_size_ segments between
+  // [begin, begin + size), returns the offset where the range begin,
+  // and the actual size unmapped (multiple of segment_size_)
+  SegmentRange unmap(SegmentRange range) {
+    auto begin = segmentRight(range.ptr);
+    auto end = segmentLeft(range.ptr + range.size);
+    if (begin >= end) {
+      return SegmentRange{range.ptr, 0};
+    }
+    unmapHandles(begin, end);
+    return rangeFromHandles(begin, end);
+  }
+
+  char* ptr() const {
+    return (char*)ptr_;
+  }
+  size_t size() const {
+    return max_handles_ * segment_size_;
+  }
+
+  void addPeer(int device) {
+    peers_.push_back(device);
+    forEachAllocatedRange(
+        [&](size_t begin, size_t end) { setAccess(device, begin, end); });
+  }
+
+  ~ExpandableSegment() {
+    forEachAllocatedRange(
+        [&](size_t begin, size_t end) { unmapHandles(begin, end); });
+    C10_CUDA_DRIVER_CHECK(cuMemAddressFree(ptr_, segment_size_ * max_handles_));
+  }
+
+ private:
+  void setAccess(int device, size_t begin, size_t end) {
+    CUmemAccessDesc desc;
+    desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    desc.location.id = device;
+    desc.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    C10_CUDA_DRIVER_CHECK(cuMemSetAccess(
+        ptr_ + begin * segment_size_, (end - begin) * segment_size_, &desc, 1));
+  }
+
+  void unmapHandles(size_t begin, size_t end) {
+    // note: unlike cudaFree, MemUnmap and MemRelease do
+    // not appear to synchronize in all cases, so we have to wait for the
+    // stream to finish before this memory is truly free.
+
+    // cannot call c10::cuda::stream_synchronize because
+    // it might grab the GIL which can lead to a deadlock
+    // Locking order must be GIL -> Allocator Lock
+    C10_CUDA_CHECK(cudaStreamSynchronize(stream_));
+    for (auto i : c10::irange(begin, end)) {
+      CUmemGenericAllocationHandle h = handles_.at(i).value();
+      handles_.at(i) = c10::nullopt;
+      C10_CUDA_DRIVER_CHECK(
+          cuMemUnmap(ptr_ + segment_size_ * i, segment_size_));
+      C10_CUDA_DRIVER_CHECK(cuMemRelease(h));
+    }
+    trimHandles();
+  }
+  void trimHandles() {
```

Comment:
nit: I don't think this is ever called in a context where we're not just clearing the whole thing

Reply:
unmapHandles calls it when unmapping only some handles.